### PR TITLE
[AIRFLOW-5268] Apply same DAG naming conventions as in literature

### PR DESCRIPTION
**Pull request by BasPH**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-5268
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

The Airflow codebase is extremely confusing because the concept "root" node in Airflow is actually implemented as the last, finishing node of a DAG, while in DAG literature root nodes are the first nodes to execute. Or, as literature also explains it: root nodes are nodes without upstream dependencies, while it was implemented as nodes without downstream dependencies.

This PR aligns the Airflow implementation of a "root" node with DAG literature. I've also implemented a `leaves` property to make a clear distinction between first/starting nodes and last/finishing nodes. Also, to my surprise there weren't even tests for these basic properties, so I added 3 tests verifying the behaviour of `roots` and `leaves`.

The implementation involved:
- Correctly implementing the definition root and leaf nodes
- Adding tests verifying this behaviour
- Some switching between upstream/downstream calls here and there because of the now correct behaviour

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

Added 3 tests, see above.

### Commits

- [ ] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [ ] Passes `flake8`

### CODE REVIEW

File: airflow/models/dag.py [hunk 1]
```diff
@@ -752,6 +752,10 @@ def get_task_instances(
 
     @property
     def roots(self):
+        return [t for t in self.tasks if not t.upstream_list]
```
Comment (Fokko) [inline airflow/models/dag.py]: while we're at it, can we use `task` instead of `t`?

File: airflow/models/dag.py [hunk 2]
```diff
@@ -752,6 +752,10 @@ def get_task_instances(
...
+
+    @property
+    def leaves(self):
```
Comment (Fokko) [inline airflow/models/dag.py]: Can we add python docs here: `Tasks that have no children. These are called leaves or leaf nodes.`

File: airflow/models/dag.py [hunk 3]
```diff
@@ -752,6 +752,10 @@ def get_task_instances(
 
     @property
     def roots(self):
```
Comment (Fokko) [inline airflow/models/dag.py]: Can we add a python doc here as well: `Tasks that have no parents. These are the roots of the DAG.`

File: airflow/models/dagrun.py [hunk 4]
```diff
@@ -308,20 +308,17 @@ def update_state(self, session=None):
         duration = (timezone.utcnow() - start_dttm).total_seconds() * 1000
         Stats.timing("dagrun.dependency-check.{}".format(self.dag_id), duration)
 
-        root_ids = [t.task_id for t in dag.roots]
-        roots = [t for t in tis if t.task_id in root_ids]
+        leaf_tis = [ti for ti in tis if ti.task_id in [t.task_id for t in dag.leaves]]
```
Comment (Fokko) [inline airflow/models/dagrun.py]:
```suggestion
        leaf_tis = [ti for ti in tis if ti.task_id in {t.task_id for t in dag.leaves}]
```

File: airflow/models/dagrun.py [hunk 5]
```diff
@@ -308,20 +308,17 @@ def update_state(self, session=None):
...
 
         # if all roots finished and at least one failed, the run failed
-        if (not unfinished_tasks and
-                any(r.state in (State.FAILED, State.UPSTREAM_FAILED) for r in roots)):
+        if not unfinished_tasks and any(l.state in (State.FAILED, State.UPSTREAM_FAILED) for l in leaf_tis):
```
Comment (Fokko) [inline airflow/models/dagrun.py]:
```suggestion
        if not unfinished_tasks and any(leaf_ti.state in {State.FAILED, State.UPSTREAM_FAILED} for leaf_ti in leaf_tis):
```

```
fokkodriesprong$ python3
Python 3.6.6 (v3.6.6:4cf1f54eb7, Jun 26 2018, 19:50:54) 
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from airflow.utils.state import State
>>> State.SUCCESS in {State.FAILED, State.UPSTREAM_FAILED}
False
>>> State.FAILED in {State.FAILED, State.UPSTREAM_FAILED}
True
```

File: airflow/models/dagrun.py [hunk 6]
```diff
@@ -308,20 +308,17 @@ def update_state(self, session=None):
...
             self.log.info('Marking run %s failed', self)
             self.set_state(State.FAILED)
             dag.handle_callback(self, success=False, reason='task_failure',
                                 session=session)
 
-        # if all roots succeeded and no unfinished tasks, the run succeeded
-        elif not unfinished_tasks and all(r.state in (State.SUCCESS, State.SKIPPED)
-                                          for r in roots):
+        # if all leafs succeeded and no unfinished tasks, the run succeeded
+        elif not unfinished_tasks and all(l.state in (State.SUCCESS, State.SKIPPED) for l in leaf_tis):
```
Comment (Fokko) [inline airflow/models/dagrun.py]:
```suggestion
        elif not unfinished_tasks and all(leaf_ti.state in {State.SUCCESS, State.SKIPPED} for leaf_ti in leaf_tis):
```

File: airflow/www/templates/airflow/graph.html [hunk 7]
```diff
@@ -142,7 +142,7 @@
 
     // Set edges
     edges.forEach(function(edge) {
-      g.setEdge(edge.u, edge.v)
+      g.setEdge(edge.source_id, edge.target_id)
```
Comment (Fokko) [inline airflow/www/templates/airflow/graph.html:145]: Nice one 👍

File: airflow/models/dag.py [hunk 8]
```diff
@@ -227,7 +227,7 @@ def __init__(
         self._description = description
         # set file location to caller source path
         self.fileloc = sys._getframe().f_back.f_code.co_filename
-        self.task_dict = dict()  # type: Dict[str, TaskInstance]
+        self.task_dict = dict()  # type: Dict[str, BaseOperator]
```
Comment (Fokko) [inline airflow/models/dag.py:230]: Is this correct?

Comment (BasPH) [reply airflow/models/dag.py:230]: Yes :( We actually fill it with Operators, not TaskInstances. The mixing of "operator" and "task" throughout the codebase is rather confusing.

Comment (BasPH) [reply airflow/models/dag.py:230]: I changed it in this PR because my IDE (PyCharm) was not suggesting the correct methods.

**Comment by BasPH** [comment]
@Fokko applied all your suggestions, fingers crossed for the CI

**Review by Fokko — APPROVED** [review]

**Review by kaxil — APPROVED** [review]
Awesome work @BasPH

**Pull request merged**